syntax = "proto3";

package printer;

import "google/protobuf/timestamp.proto";

message PrintDocument {
  string document_id = 1;
  string name = 2;
  bytes content = 3;
  PrintSettings settings = 4;
}

message PrintSettings {
  enum ColorMode {
    COLOR_UNSPECIFIED = 0;
    COLOR = 1;
    GRAYSCALE = 2;
  }

  enum PaperSize {
    SIZE_UNSPECIFIED = 0;
    A4 = 1;
    LETTER = 2;
    LEGAL = 3;
  }

  enum Orientation {
    ORIENTATION_UNSPECIFIED = 0;
    PORTRAIT = 1;
    LANDSCAPE = 2;
  }

  ColorMode color_mode = 1;
  PaperSize paper_size = 2;
  Orientation orientation = 3;
  int32 copies = 4;
  bool double_sided = 5;
}

message PrintJob {
  string job_id = 1;
  string document_id = 2;
  google.protobuf.Timestamp submitted_at = 3;
  JobStatus status = 4;
  int32 pages_printed = 5;
  int32 total_pages = 6;
}

enum JobStatus {
  JOB_STATUS_UNSPECIFIED = 0;
  PENDING = 1;
  PROCESSING = 2;
  COMPLETED = 3;
  FAILED = 4;
  CANCELLED = 5;
}

message PrinterStatus {
  enum State {
    STATE_UNSPECIFIED = 0;
    READY = 1;
    BUSY = 2;
    ERROR = 3;
    OFFLINE = 4;
  }

  State state = 1;
  string error_message = 2;
  int32 jobs_in_queue = 3;
  bool paper_jam = 4;
  int32 paper_level = 5;
  int32 toner_level = 6;
}